p2p/pkg/stakemanager/stakemanager.go

```diff
@@ -0,0 +1,308 @@
+	"github.com/primev/mev-commit/x/contracts/events"
+	"golang.org/x/sync/errgroup"
+)
+
+type ProviderRegistryContract interface {
+	GetProviderStake(*bind.CallOpts, common.Address) (*big.Int, error)
+	MinStake(*bind.CallOpts) (*big.Int, error)
+}
+
+type StakeManager struct {
+	owner            common.Address
+	evtMgr           events.EventManager
+	providerRegistry ProviderRegistryContract
+	notifier         notifications.Notifier
+	stakeMu          sync.RWMutex
+	stakes           *lru.Cache[common.Address, *big.Int]
+	unstakeReqs      *lru.Cache[common.Address, struct{}]
+	minStake         atomic.Pointer[big.Int]
+	logger           *slog.Logger
+}
+
+func NewStakeManager(
+	owner common.Address,
+	evtMgr events.EventManager,
+	providerRegistry ProviderRegistryContract,
+	notifier notifications.Notifier,
+	logger *slog.Logger,
+) (*StakeManager, error) {
+	minStake, err := providerRegistry.MinStake(&bind.CallOpts{
+		From: owner,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get min stake: %w", err)
+	}
+	stakes, err := lru.New[common.Address, *big.Int](1000)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create stakes cache: %w", err)
+	}
+	unstakeReqs, err := lru.New[common.Address, struct{}](1000)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create unstake requests cache: %w", err)
+	}
+	var minStakePtr atomic.Pointer[big.Int]
+	minStakePtr.Store(minStake)
+	return &StakeManager{
+		providerRegistry: providerRegistry,
+		evtMgr:           evtMgr,
+		logger:           logger,
+		owner:            owner,
+		stakes:           stakes,
+		unstakeReqs:      unstakeReqs,
+		minStake:         minStakePtr,
+		notifier:         notifier,
+	}, nil
+}
+
+func (sm *StakeManager) Start(ctx context.Context) <-chan struct{} {
+	doneChan := make(chan struct{})
+
+	eg, egCtx := errgroup.WithContext(ctx)
+
+	ch1 := make(chan *providerregistry.ProviderregistryProviderRegistered)
+	ev1 := events.NewChannelEventHandler(egCtx, "ProviderRegistered", ch1)
+
+	ch2 := make(chan *providerregistry.ProviderregistryFundsSlashed)
+	ev2 := events.NewChannelEventHandler(egCtx, "FundsSlashed", ch2)
+
+	ch3 := make(chan *providerregistry.ProviderregistryFundsDeposited)
+	ev3 := events.NewChannelEventHandler(egCtx, "FundsDeposited", ch3)
+
+	ch4 := make(chan *providerregistry.ProviderregistryUnstake)
+	ev4 := events.NewChannelEventHandler(egCtx, "Unstake", ch4)
+
+	ch5 := make(chan *providerregistry.ProviderregistryMinStakeUpdated)
+	ev5 := events.NewChannelEventHandler(egCtx, "MinStakeUpdated", ch5)
+
+	sub, err := sm.evtMgr.Subscribe(ev1, ev2, ev3, ev4, ev5)
+	if err != nil {
+		close(doneChan)
+		return doneChan
+	}
+
+	eg.Go(func() error {
+		defer sub.Unsubscribe()
+
+		select {
+		case <-egCtx.Done():
+			sm.logger.Info("event subscription context done")
+			return nil
+		case err := <-sub.Err():
+			return fmt.Errorf("error in event subscription: %w", err)
+		}
+	})
+
+	eg.Go(func() error {
```

Comment:
Why do we handle all the events in a single goroutine? Probably FundsSlashed event will be more frequent, than other events

Reply:
I did not want to add more routines. I had planned to use buffered channels. Fixed it.